### DISCUSSION TITLE
Unable to prevent dependency upgrade using postUpgradeTasks

### DISCUSSION THREAD

**Discussion by ondrejbenes** [renovatebot/renovate#15350] [Request Help]
Hello,
I would like to avoid upgrade of a dependency in a lockfile (requirements.txt) when it is not present in the main dependencies file (requirements.in).

I have the following config:
```
    "packageRules": [
        {
            "matchManagers": [
                "pip_requirements"
            ],
            "postUpgradeTasks": {
                "commands": [
                    "cd $(dirname {{{packageFile}}}) && sed -i 's/{{{depName}}}=={{{newVersion}}}/{{{depName}}}=={{{currentVersion}}}/' {{{packageFile}}} && (pip-compile --upgrade-package={{{depName}}}=={{{newVersion}}} $(basename {{{packageFile}}} .txt).in || true)"
                ],
                "fileFilters": [
                    "**/requirements*.txt"
                ]
            }
        }
    ],
 ```

Even though the `sed` command rollbacks the upgrade (checked using `cat {{{packageFile}}}` after `pip-compile`), the version is bumped anyway (I assume in a followup step).

Strangely enough, when I modify the `packageFile` (i.e. `echo hello >> {{{packageFile}}}`), the version is not upgraded (as expected), but this does not solve my issue, because the commit is made, just with a different change.

Note: I tried using `pip-compile` manager, but it does not work.

**Comment by PhilipAbed** [comment]
I'm trying to understand the issue,
You use pip-compile package manager.
It has a direct dependencies file called `requirements.in`
and the usual pip file `requirements.txt` which has both Direct dependencies and Transitive dependencies.
and you want to only update dependencies which appear in `requirements.in`?

just for clarification: you do want changes to happen in `requirements.txt` but only for direct dependencies which also appear in `requiremnets.in` right??

I believe it is possible to ignore the requirements.txt file completely, but not sure renovate supports anyway to get part of the dependencies in the `requirements.txt` that exist in `requirements.in`

the files are checked individually without connection since `requirements.txt` is also a `pip` file which is considered a package manager,
and the `requirements.in` can be in a different folder than the `requirements.txt`, there's no clear way as far as i know to identify the connection between the `requirements.in` and `requirements.txt` files.

pip-compile is still considered in an Alpha Stage, so any feature ideas are welcome, but you will have to provide information on how to do it.

An idea would be for example to create a Map for pip-compile , which will have the path to `requirements.in` as key and the path to `requirements.txt` as value, that's one way to connect them through config and do what you want.

if that is what you want, maybe we can open a feature request
If any of what i said above is wrong, or i misunderstood, please fix me

**Reply by ondrejbenes** [reply]
I should have specified why I consider `pip-compile` manager not to work. The issue is that it seems to require version pinning in the `requirements.in` file, which seems quite impractical to me.

> just for clarification: you do want changes to happen in requirements.txt but only for direct dependencies which also appear in requiremnets.in right??

yes

> pip-compile is still considered in an Alpha Stage, so any feature ideas are welcome, but you will have to provide information on how to do it.

I suppose I can create a feature request to support unpinned dependencies in the `requirements.in` file. Unless you can think of a reason why that would be a bad idea?

**Reply by viceice** [reply]
Why is pinning impractical? That would solve your use-case.

**Reply by viceice** [reply]
We can support unpinned deps, but it needs to extract the current version from txt file as `lockedVersion` so the `pip-compile` manager can support `rangeStrategy=update-lockfile`

https://docs.renovatebot.com/configuration-options/#rangestrategy

**Reply by viceice** [reply]
https://docs.renovatebot.com/dependency-pinning/ this is also parially valid for all other managers

**Reply by ondrejbenes** [reply]
Three reasons I suppose:

1) addition of a new dependency now requires multiple steps - you need to first add it to `requirements.in`, then run pip-compile to see to which version the dependency is resolved, then you need to actually pin it in `requirements.in`
2) we already have a number of repositories with requirements.in, pinning the versions in each one would require a lot of effort
3) there would need to be some sort of mechanism (e.g. pre-commit hook, CI lint job) in all repositories that would verify that all dependencies in requirements.in are pinned (otherwise, renovatebot will not update them)

Reason 3 is the biggest blocker, as it requires a long term solution that will need to be maintained.

**Comment by viceice** [comment]
i think it's a bad idea to block transitive Updates from pip-compile as it can break the whole app.

please see our docs repo for pip-compile usage. we only update the `in` file from renovate and let pip-compile update the `txt` file as required.

https://github.com/renovatebot/renovatebot.github.io/blob/main/renovate.json

**Reply by ondrejbenes** [reply]
I'm perfectly fine with updates of transitive dependencies when it happens as part of `pip-compile` command. However, I am using the regular `pip_requirements`, not `pip-compile` manager (see above why not).

**Comment by PhilipAbed** [comment]
@ondrejbenes i think it's not something that is easy to do because
lets say we have Dependency `A(1.0.0)` , `A(1.0.0)` has 1 transitive `B(1.0.0)`
now if we upgrade `A(2.0.0)`, it might have more transitive deps, lets say it has 3 transitive deps now after releasing new version.

you have to run a command to update the lock file accordingly.
i'm with Viceice on this.
